File: Server_Root/Server/Server/JoinLobbyToS.cpp
#include "stdafx.h"
#include "JoinLobbyToS.h"
#include "NetworkDefines.h"
#include "JSONHelper.h"


JoinLobbyToS::JoinLobbyToS()
{
	m_buffer = new GenericStringBuffer<UTF8<>>();
	m_writer = new Writer<StringBuffer, UTF8<>>(*m_buffer);
}

JoinLobbyToS::~JoinLobbyToS()
{
	delete m_buffer;
	delete m_writer;
}

unsigned short JoinLobbyToS::GetID()
{
	return (unsigned short)Messages::JoinLobbyToS_ID;
}

const char* JoinLobbyToS::Serialize()
{
	Document document;
	document.SetObject();

	JSONHelper::AddField(&document, "PlayerKey", m_strPlayerKey);
	JSONHelper::AddField(&document, "AuthKey", m_nAuthKey);

	m_buffer->Clear();
	document.Accept(*m_writer);

	return m_buffer->GetString();
}

bool JoinLobbyToS::Deserialize(const char* pChar)
{
	Document document;
	document.Parse<0>(pChar);
	if (!document.IsObject())
	{
		return false;
	}

	if (!JSONHelper::GetField(&document, "PlayerKey", &m_strPlayerKey)) return false;
	if (!JSONHelper::GetField(&document, "AuthKey", &m_nAuthKey)) return false;

	return true;
}

File: Client_Root/Client/Assets/Scripts/Network/Messages/ToServer/JoinLobbyToS.cs
using System.Text;

public class JoinLobbyToS : IMessage
{
    public string   m_strPlayerKey;         //  json field name : PlayerKey
    public int      m_nAuthKey;             //  json field name : AuthKey

    public ushort GetID()
    {
        return (ushort)Messages.JoinLobbyToS_ID;
    }

    public byte[] Serialize()
    {
        JSONObject jsonObj = new JSONObject(JSONObject.Type.OBJECT);

        JSONHelper.AddField(jsonObj, "PlayerKey", m_strPlayerKey);
        JSONHelper.AddField(jsonObj, "AuthKey", m_nAuthKey);

        return Encoding.Default.GetBytes(jsonObj.Print());
    }

    public bool Deserialize(byte[] bytes)
    {
        JSONObject jsonObj = new JSONObject(Encoding.UTF8.GetString(bytes));

        if(!JSONHelper.GetField(jsonObj, "PlayerNumber", ref m_strPlayerKey)) return false;
        if(!JSONHelper.GetField(jsonObj, "AuthKey", ref m_nAuthKey)) return false;

        return true;
    }
}
